### ISSUE TITLE
Laravel Driver - % LIKE % in Primary Key 

### ISSUE THREAD

**Issue by vertisan** [teamtnt/laravel-scout-tntsearch-driver#78]
I'm using Laravel Scout with TNTSearch Engine at it's works fine but with one little problem.
 I have records:
```
| ID      | Name     |
+---------+----------+
| 9030100 | Car      |
| 9030150 | Car2     |
| 9030200 | Radio    |
```
And here is my query:
`CatalogProducts::search( $query )->paginate( 15 );`

When I'm looking '**car**', it's returning **all records with 'car' in name**

When I'm looking '**9030100**', it's **returning single product 'Car'**

But when I'm looking for '**9030**' **I don't have any results** and here is question - why?

Where is the problem? 

P.S. ID is string and is Primary Key

**Comment by nticaric** [comment]
TNTSearch isn't a replace for the LIKE operator, however you would get the wanted result if you enable fuzziness

**Comment by nticaric** [comment]
Another approach is to use n-gram. Take a look at this demo 

http://cities.tnt.studio/

**Comment by pollocknoah** [comment]
Hi vertisan I think it's to do with your TNT settings in `config/scout.php`

In the Readme they have a recommended value but it doesn't seem to be a good default.

Changing them to this may give you better results : 

```
    'tntsearch' => [
    'storage' => storage_path(),
        'fuzziness' => env('TNTSEARCH_FUZZINESS', true),
        'fuzzy' => [
            'prefix_length' => 2, //The number of initial characters which will not be “fuzzified”. This helps to reduce the number of terms which must be examined. Defaults to 0.
            'max_expansions' => 50, //The maximum number of terms that the fuzzy query will expand to. Defaults to 50.
            'distance' => 20, // I don't fully get this one, but appears to be the length of characters it extends to, a previous explanation says its the Levenshtein Distance, but it doesn't seem to do that in my case. The fuzziness setting seems to describe the Levenshtein Distance.
        ],
    'asYouType' => false
    ],
```

and in your `.env` file include: `TNTSEARCH_FUZZINESS=auto`

The `fuzziness` setting is described in this doc : 
https://www.elastic.co/guide/en/elasticsearch/reference/current/common-options.html#fuzziness

**TL:DR**
The fuzziness parameter can be specified as:

`0, 1, 2`
the maximum allowed Levenshtein Edit Distance (or number of edits)
`AUTO`
generates an edit distance based on the length of the term. For lengths:
`0..2`
must match exactly
`3..5`
one edit allowed
`>5`
two edits allowed

`AUTO` should generally be the preferred value for fuzziness.

This will return more results and your examples should show up.

**Comment by nticaric** [comment]
@pollocknoah let me try to explain the fuzziness a little bit better. Forget that we're using a search package, just imagine a regular table in your database. Let's say it has the following records:

ID | City
------------ | -------------
1|Amsterdam
2|Antwerp
3|Athens
4|Barcelona
5|Belgrade
6|Berlin
7|Bilbao
8|Birmingham
9|Bremen
10|Budapest
11|Copenhagen
12|Dublin
13|Frankfurt
14|Hamburg

Now if you do a search like this `%min%` it will return only one row, matching `Birmingham`, but the
internals of your database will go through each record to check if the city contains a "%min%". So if you have a large record set the complexity of this query is O(N) which is bad. Can we do better? Yes!

Much better performance can be get if the wildcard operator is only on the right side, not on both sides, something like this `bir%`. This would leverage the btree index of the city column, so the complexity is no longer O(N) but O(log n) which is a huge difference in a large dataset.

Lets get back to our example. You want to search for "Brem" expecting the result set to contain the city Bremen, right. Here's what's happening if the fuzziness is enabled. The `prefix_length` is by default set to 2 so we'll extract the first 2 letters from your query, which is "br". Then, we ask the database

`select * from cities where city like "br%"`

From our example above, this will match only one record, but if we had all cities in the world in our table, this would return a lot of results, but still with good performance, because the complexity is
 O(log n). We didn't have to go through, all records, but only through the "br%" branch. 
Let's say we had all of the cities, how many records do we return starting with "br%". By default we return 50, because thats the `max_expansions` setting, so we save some memory and improve performance. We don't return just any 50 records that start with br but the most frequent one.

Once we have those 50 records returned we calculate the levenshtein distance between each result and your query. So your query was "Brem" and the result record was "Bremen". Since we need to append to characters the levenshtein distance is 2, so this record will be returned, if the distance, was for example 3, it wouldn't be returned.

Back to your case, setting the levenshtein distance to 20 is like ignoring it. Fuzziness in TNTSearch isn't meant to replace the "LIKE" operator but to spot some basic typos, for which 2 works best

Hope this helps a little bit to understand the fuzziness. Feel free to ask any questions :)

**Comment by vertisan** [comment]
Thank you very much for your explain, now I know everythink ;)

**Comment by pollocknoah** [comment]
@nticaric Thanks for your explanation. It's helping me understand it better. I'm implementing a search function for work and by setting the `distance` to `20` it felt like I wasn't doing something write.

I still needs a few points cleared up in terms of your recommended settings.
1. On the line : `'fuzziness' => env('TNTSEARCH_FUZZINESS', false)` in `config/scouts.php. What is the value you recommending setting it at in the `.env` if any? Is `true` or `false` your default? Since `false` is in your documentation is that what you would recommend?

2. Does the `fuzziness` setting have anything to do with the `'fuzzy' => [ 'distance' : <number> ]` number set here at all ? If you have one do you need the other ?

3. To follow up on your example of `Breman` I don't seem to be always getting same behaviour.
I will demonstrate with my settings and the model I am searching for. I won't list you my Customers dB table, but the results that show up hopefully will be enough.

The only parameter I will change will be `distance` all the other ones I'll keep the same.
Here are my settings :
```
    'tntsearch' => [
    'storage' => storage_path(),
        'fuzziness' => env('TNTSEARCH_FUZZINESS', false),
        'fuzzy' => [
            'prefix_length' => 2,
            'max_expansions' => 50, 
            'distance' => CHANGES FOR EACH EXAMPLE BELOW,
        ],
    'asYouType' => false
    ],
```
`TNTSEARCH_FUZZINESS=auto` in `.env`

This is an example of an entry I'm trying to search for : 
```
App\Customer {#849
         id: 10,
         partner_id: 8,
         customer_status_id: 8,
         first_name: "Justine",
         last_name: "Cremin",
         email: "torrance.schuppe@white.com",
         telephone: "748-438-0112",
         created_at: "2017-03-31 14:54:51",
         updated_at: "2017-03-31 14:54:51",
       },
```

The results are done through `php artisan tinker` and each time the distance setting is changed I `dump-autoload` and restart `tinker`.

*Results :*

**Example 1:**

`'distance' => '2'` 

```
>>> Customer::search('Jus')->get();
=> Illuminate\Database\Eloquent\Collection {#768
     all: [],
   }
>>> Customer::search('Just')->get();
=> Illuminate\Database\Eloquent\Collection {#797
     all: [
       App\Customer {#844
         id: 10,
         partner_id: 8,
         customer_status_id: 8,
         first_name: "Justine",
         last_name: "Cremin",
         email: "torrance.schuppe@white.com",
         telephone: "748-438-0112",
         created_at: "2017-03-31 14:54:51",
         updated_at: "2017-03-31 14:54:51",
       },
     ],
   }
```

**Q:** If the distance is 2 should this not show up as it isn't enough to complete the 'e' ? Or does the search return the closest result?

**Example 2:**

`'distance' => '4'`

```
>>> Customer::search('Ju')->get();
=> Illuminate\Database\Eloquent\Collection {#794
     all: [
       App\Customer {#855
         id: 10,
         partner_id: 8,
         customer_status_id: 8,
         first_name: "Justine",
         last_name: "Cremin",
         email: "torrance.schuppe@white.com",
         telephone: "748-438-0112",
         created_at: "2017-03-31 14:54:51",
         updated_at: "2017-03-31 14:54:51",
       },
     ],
   }
>>> Customer::search('Justi')->get();
=> Illuminate\Database\Eloquent\Collection {#801
     all: [
       App\Customer {#802
         id: 10,
         partner_id: 8,
         customer_status_id: 8,
         first_name: "Justine",
         last_name: "Cremin",
         email: "torrance.schuppe@white.com",
         telephone: "748-438-0112",
         created_at: "2017-03-31 14:54:51",
         updated_at: "2017-03-31 14:54:51",
       },
     ],
   }

```

**Q:** If the distance is 4 and I'm searching for 'Ju', it's one character short of 'Justine'. The second search for 'Justi' with a distance of 4 goes over her name, but the result still comes up.


**Q:** What sort of setting is best for a general customer search? When someone types in the search field results similar to what they type will start to show up. This could be any parameter, their first name, email, telephone. Is there such a setting that does this without putting too much strain on the system.

ie. 

Customer to search for : 
```
App\Customer {#770
         id: 37,
         partner_id: 35,
         customer_status_id: 35,
         first_name: "Edd",
         last_name: "Graham",
         email: "walter.jake@schneider.com",
         telephone: "604-925-1276 x03200",
         created_at: "2017-04-03 11:10:22",
         updated_at: "2017-04-03 11:10:22",
       },
```

When a person types in the search bar `schn` I would like that query to be enough for that model to come up based on their email.

At `'distance' => '2'` nothing comes up.
```
>>> Customer::search('schn')->get();
=> Illuminate\Database\Eloquent\Collection {#857
     all: [],
   }
```
  
At `'distance' => '4'` 2 other models come up, BUT not the one I want. Still unclear as to why these are the results for this setting: 

```
>>> Customer::search('schn')->get();
=> Illuminate\Database\Eloquent\Collection {#799
     all: [
       App\Customer {#848
         id: 10,
         partner_id: 8,
         customer_status_id: 8,
         first_name: "Justine",
         last_name: "Cremin",
         email: "torrance.schuppe@white.com",
         telephone: "748-438-0112",
         created_at: "2017-03-31 14:54:51",
         updated_at: "2017-03-31 14:54:51",
       },
       App\Customer {#849
         id: 19,
         partner_id: 17,
         customer_status_id: 17,
         first_name: "Kaleb",
         last_name: "Schultz",
         email: "imills@gmail.com",
         telephone: "1-961-657-2608 x57411",
         created_at: "2017-03-31 15:24:25",
         updated_at: "2017-03-31 15:24:25",
       },
     ],
   }
```

At `'distance' => '6'` MANY model comes up INCLUDING the one I'm looking for 

```
>>> Customer::search('schn')->get();
=> Illuminate\Database\Eloquent\Collection {#799
     all: [
       App\Customer {#851
         id: 6,
         partner_id: 4,
         customer_status_id: 4,
         first_name: "narp",
         last_name: "Schulist",
         email: "mittie.schulist@yahoo.com",
         telephone: "(320) 4989 -6779",
         created_at: "2017-03-31 14:54:51",
         updated_at: "2017-03-31 14:54:51",
       },
       App\Customer {#852
         id: 10,
         partner_id: 8,
         customer_status_id: 8,
         first_name: "Justine",
         last_name: "Cremin",
         email: "torrance.schuppe@white.com",
         telephone: "748-438-0112",
         created_at: "2017-03-31 14:54:51",
         updated_at: "2017-03-31 14:54:51",
       },
       App\Customer {#853
         id: 19,
         partner_id: 17,
         customer_status_id: 17,
         first_name: "Kaleb",
         last_name: "Schultz",
         email: "imills@gmail.com",
         telephone: "1-961-657-2608 x57411",
         created_at: "2017-03-31 15:24:25",
         updated_at: "2017-03-31 15:24:25",
       },
       App\Customer {#854
         id: 21,
         partner_id: 19,
         customer_status_id: 19,
         first_name: "Kayleigh",
         last_name: "Beer",
         email: "della.kovacek@schuster.info",
         telephone: "486.257.6669 x162",
         created_at: "2017-03-31 15:24:25",
         updated_at: "2017-03-31 15:24:25",
       },
       App\Customer {#855
         id: 37,
         partner_id: 35,
         customer_status_id: 35,
         first_name: "Edd",
         last_name: "Graham",
         email: "walter.jake@schneider.com",
         telephone: "604-925-1276 x03200",
         created_at: "2017-04-03 11:10:22",
         updated_at: "2017-04-03 11:10:22",
       },
     ],
   }
```

With this setting it doesn't seem narrow it to the SINGLE object until I type it in correctly and is still not quite what I'm looking to achieve.

Is this where your recommendation of n-grams come in?
Is there a way to achieve an interactive search bar with this package or do I need to combine what is provided here with the n-grams tutorial?

**Comment by nticaric** [comment]
Great questions! Hopefully I'll be able to explain them.  Some basic linguistics and information retrieval theory is required.

The behavior that you're experiencing is because of the stemming process. The index actually doesn't contain the real words but a stemmed version of them, which we call terms. To put it simple, if you have the word `running`it will be stemmed to `run`, if you have the word `lazy` it will be stemmed to `lazi`. It tries to find a common root of each word based on a stemming algorithm. The default one is Porter stemmer. The stemming process is necessary in order to give you correct results when you don't know the right word form upfront. So if you search for the word `consolidate` it will return documents that have `consolidate`, `consolidated` or `consolidating` in it because the stemmed version of all of the is `consolid`. So far so good, back to your example.

The word `Justine` is actually saved in the index as `Justin` which is the stemmed version.

**Example 1**
`'distance' => '2'`
When you search for `Jus` the levenshtein distance between `Justin` and `Jus` is 3 so you get nothing.
When you search for `Just` the levenshtein distance between `Justin` and `Just` is 2 so you get one result.

**Example 2**
`'distance' => '4'`
When you search for `Ju` the levenshtein distance between `Justin` and `Ju` is 4 so you get a result.
When you search for `Justi` the levenshtein distance between `Justin` and `Just` is 1 so you get one result.

For searching a customer table where the names don't come in different word forms it's good to turn off the stemming process. The laravel scout driver is somehow limited to do this easily so what I would suggest is using the TNTSearch package directly. There is also an option called `asYouType`
that tries to guess a best match as you type. So this might fit your needs.

So my first recommendation is:
* turn of the stemmer
* enable asYouType

My second and ultimate suggestion that will work like magic is:
* turn of the stemmer
* build n-grams (3 grams in particular for english) for each customer
* break your search query into 3-grams and query the n-gram index

http://cities.tnt.studio/ is a good example of how your customer search could work. So if there is an exact match it will return it, otherwise it will give you some suggestions that are the closest

Here's the code
https://github.com/teamtnt/cities/blob/master/app/Http/Controllers/CityController.php.php#L13

**Comment by pollocknoah** [comment]
@nticaric Thanks so much for the detailed explanation. I had no idea about Stemming and I've had a look through the Stemmer folder in TNTSearch and it's making a lot more sense now.

**Q**: Can you confirm that's the right setting for TNTSEARCH_FUZZINESS ?
in my `.env` `TNTSEARCH_FUZZINESS=auto`

**Q**: I was also unable to find a place to turn off the stemming. Where would such a setting go?
I found in `tntsearch/src/TNTSearch.php` there is a function to setStemmer and if it doesn't find one in the index it defaults to a `PorterStemmer`

line 314 in `tntsearch/src/TNTSearch.php`

Is there a place in config to set the desired Stemmer somewhere in the config?

Thanks again!
